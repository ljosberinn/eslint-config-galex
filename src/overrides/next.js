/* eslint-disable inclusive-language/use-inclusive-words */
const files = ['src/pages/*.?(js|ts)?(x)', 'pages/*.?(js|ts)?(x)'];

const createNextJsOverride = ({
  react,
  rules: customRules = {},
  files: customFiles = [],
}) => {
  if (!react.isNext) {
    return null;
  }

  const rules = {
    ...defaultNextJsRules,
    ...customRules,
  };

  const finalFiles = [...new Set([...files, ...customFiles])];

  return {
    files: finalFiles,
    rules,
  };
};

const defaultNextJsRules = {
  /**
   * off because Next.js page routing requires default exports
   */
  'import/no-default-export': 'off',

  /**
   * ensures core `a` attributes are valid
   *
   * exclude `noHref` validation for Nextjs because
   *
   * @example
   * ```js
   * <Link passHref href="/foo">
   *  <a>hi</a
   * </Link>
   * ```
   * is valid
   *
   * @see https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/anchor-is-valid.md
   */
  'jsx-a11y/anchor-is-valid': [
    'error',
    {
      aspects: ['invalidHref'],
      components: ['Link'],
    },
  ],

  'react/react-in-jsx-scope': 'off',

  /**
   * ensures stylesheets are preloaded
   *
   * @see https://github.com/vercel/next.js/blob/canary/packages/eslint-plugin-next/lib/rules/missing-preload.js
   */
  '@next/next/missing-preload': 'warn',

  /**
   * should be imported directly
   *
   * @see https://github.com/vercel/next.js/blob/canary/packages/eslint-plugin-next/lib/rules/no-css-tags.js
   */
  '@next/next/no-css-tags': 'warn',

  /**
   * disallows regular <a> links
   *
   * @see https://github.com/vercel/next.js/blob/canary/packages/eslint-plugin-next/lib/rules/no-html-link-for-pages.js
   */
  '@next/next/no-html-link-for-pages': 'warn',

  /**
   * sync scripts can impact performance
   *
   * @see https://github.com/vercel/next.js/blob/canary/packages/eslint-plugin-next/lib/rules/no-sync-scripts.js
   */
  '@next/next/no-sync-scripts': 'warn',

  /**
   * disallow of polyfill.io in some cases
   *
   * @see https://github.com/vercel/next.js/blob/canary/packages/eslint-plugin-next/lib/rules/no-unwanted-polyfillio.js
   */
  '@next/next/no-unwanted-polyfillio': 'warn',
};

module.exports = {
  createNextJsOverride,
  defaultNextJsRules,
};
