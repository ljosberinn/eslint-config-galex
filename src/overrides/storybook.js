const {
  storybookOverrideType: overrideType,
} = require('../utils/overrideTypes');

const files = ['*.stories.*'];

/**
 * @param {{
 *  hasStorybook: boolean;
 *  rules?: Record<string, any>;
 * files?: string[];
 * }} options
 */
const createStorybookOverride = ({
  hasStorybook,
  rules: customRules = {},
  files: customFiles = [],
}) => {
  if (!hasStorybook) {
    return null;
  }

  const finalRules = {
    ...rules,
    ...customRules,
  };

  const finalFiles = [...new Set([...customFiles, ...files])];

  return {
    rules: finalRules,
    files: finalFiles,
    overrideType,
  };
};

const rules = {
  'import/no-default-export': 'off',
  /**
   * allow exporting anonymous config object
   */
  'import/no-anonymous-default-export': 'off',
  /**
   * allow typing template components with Storybook's `Story` type
   */
  'react/function-component-definition': 'off',
};

module.exports = {
  createStorybookOverride,
  files,
  rules,
  overrideType,
};
