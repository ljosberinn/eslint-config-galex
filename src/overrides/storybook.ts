import { OverrideCreator, RulesCreator } from '../types';
import { uniqueArrayEntries } from '../utils/array';
import { storybookOverrideType } from '../utils/overrideType';

const files = ['*.stories.*', '*.story.*'];
const plugins = ['storybook'];

export const createStorybookOverride: OverrideCreator = ({
  rules: customRules = {},
  files: customFiles,
  env: customEnv,
  excludedFiles: customExcludedFiles,
  globals: customGlobals,
  extends: customExtends,
  parserOptions: customParserOptions,
  plugins: customPlugins,
  overrides: customOverrides,
  settings: customSettings,
  ...dependencies
}) => {
  if (!dependencies.storybook.hasStorybook) {
    return null;
  }

  const finalRules = {
    ...createStorybookRules(dependencies),
    ...customRules,
  };

  const finalFiles = customFiles || files;
  const finalPlugins = uniqueArrayEntries([
    ...plugins,
    ...(customPlugins ?? []),
  ]);

  return {
    rules: finalRules,
    files: finalFiles,
    plugins: finalPlugins,
    overrideType: storybookOverrideType,
    env: customEnv,
    globals: customGlobals,
    excludedFiles: customExcludedFiles,
    extends: customExtends,
    overrides: customOverrides,
    parserOptions: customParserOptions,
    settings: customSettings,
  };
};

export const createStorybookRules: RulesCreator = ({
  react: { hasReact },
  typescript: { hasTypeScript },
  storybook: { hasStorybookTestingLibrary },
}) => {
  return {
    /**
     * @see https://github.com/storybookjs/eslint-plugin-storybook/blob/main/docs/rules/await-interactions.md
     *
     * storybook/testing-library
     */
    ...(hasStorybookTestingLibrary
      ? { 'storybook/await-interactions': 'error' }
      : null),

    /**
     * @see https://github.com/storybookjs/eslint-plugin-storybook/blob/main/docs/rules/context-in-play-function.md
     *
     * storybook/testing-library
     */
    ...(hasStorybookTestingLibrary
      ? { 'storybook/context-in-play-function': 'error' }
      : null),

    /**
     * enforces setting `component` property in unnamed default export object of a story
     *
     * @see https://github.com/storybookjs/eslint-plugin-storybook/blob/main/docs/rules/csf-component.md
     */
    'storybook/csf-component': 'error',

    /**
     * enforces having a default export.
     *
     * disabled because youre literally not seeing anything if you don't have one
     *
     * @see https://github.com/storybookjs/eslint-plugin-storybook/blob/main/docs/rules/default-exports.md
     *
     */
    'storybook/default-exports': 'off',

    /**
     * suggests modern/correct hierarchy separator
     *
     * @see https://github.com/storybookjs/eslint-plugin-storybook/blob/main/docs/rules/hierarchy-separator.md
     */
    'storybook/hierarchy-separator': 'warn',

    /**
     * suggests inlining meta properties of stories to the default export
     *
     * @see https://github.com/storybookjs/eslint-plugin-storybook/blob/main/docs/rules/meta-inline-properties.md
     */
    'storybook/meta-inline-properties': 'warn',

    /**
     * prevents redundant duplicate info
     *
     * @see https://github.com/storybookjs/eslint-plugin-storybook/blob/main/docs/rules/no-redundant-story-name.md
     */
    'storybook/no-redundant-story-name': 'warn',

    /**
     * prevents using the deprecated `storiesOf` api
     *
     * @see https://github.com/storybookjs/eslint-plugin-storybook/blob/main/docs/rules/no-stories-of.md
     */
    'storybook/no-stories-of': 'error',

    /**
     * removes redundant title property of unnamed default exports in storybook
     * as its inferred from path
     *
     * @see https://github.com/storybookjs/eslint-plugin-storybook/blob/main/docs/rules/no-title-property-in-meta.md
     */
    'storybook/no-title-property-in-meta': 'warn',

    /**
     * prefer PascalCase for named exports
     *
     * @see https://github.com/storybookjs/eslint-plugin-storybook/blob/main/docs/rules/prefer-pascal-case.md
     */
    'storybook/prefer-pascal-case': 'warn',

    /**
     * require at least one named export
     *
     * @see https://github.com/storybookjs/eslint-plugin-storybook/blob/main/docs/rules/story-exports.md
     */
    'storybook/story-exports': 'warn',

    /**
     * requires importing `expect` from `@storybook/jest`
     *
     * @see https://github.com/storybookjs/eslint-plugin-storybook/blob/main/docs/rules/use-storybook-expect.md
     */
    'storybook/use-storybook-expect': 'error',

    /**
     * use `@storybook/testing-library` wrapper over `@testing-library/react` for
     * interactions
     *
     * active even if the wrapper is not present to raise awareness
     *
     * @see https://github.com/storybookjs/eslint-plugin-storybook/blob/main/docs/rules/use-storybook-testing-library.md
     */
    'storybook/use-storybook-testing-library': 'error',

    /**
     * default exports are mandatory in storybook
     */
    'import/no-default-export': 'off',

    /**
     * allow exporting anonymous config object
     */
    'import/no-anonymous-default-export': 'off',

    /**
     * allow typing template components with Storybook's `Story` type
     */
    ...(hasReact ? { 'react/function-component-definition': 'off' } : null),

    /**
     * disable return type requirement
     */
    ...(hasTypeScript
      ? { '@typescript-eslint/explicit-module-boundary-types': 'off' }
      : null),

    /**
     * allow dummy functions satisfying type requirements
     */
    ...(hasTypeScript
      ? { '@typescript-eslint/no-empty-function': 'off' }
      : null),
  };
};
