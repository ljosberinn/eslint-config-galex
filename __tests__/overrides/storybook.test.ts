import { createStorybookOverride } from '../../src/overrides/storybook';
import type { OverrideCreator } from '../../src/types';
import { defaultProject } from '../shared';

const storybookDefaultProject: Parameters<OverrideCreator>[0] = {
  ...defaultProject,
  storybook: {
    ...defaultProject.storybook,
    hasStorybook: true,
  },
};

test('allows passing rules', () => {
  const ruleName = 'foo';
  const ruleValue = 'warn';

  const result = createStorybookOverride({
    ...storybookDefaultProject,
    rules: {
      [ruleName]: ruleValue,
    },
  });

  expect(result?.rules[ruleName]).toBe(ruleValue);
});

test('allows passing custom files', () => {
  const files = ['foo/bar'];

  const result = createStorybookOverride({
    ...storybookDefaultProject,
    files,
  });

  expect(result?.files).toBe(files);
});

test('allows passing parserOptions', () => {
  const parserOptions: Parameters<OverrideCreator>[0]['parserOptions'] = {
    ecmaVersion: 2017,
  };

  const result = createStorybookOverride({
    ...storybookDefaultProject,
    parserOptions,
  });

  expect(result?.parserOptions?.ecmaVersion).toStrictEqual(
    parserOptions.ecmaVersion
  );
});

test('allows passing settings', () => {
  const settings: Parameters<OverrideCreator>[0]['settings'] = {
    foo: 'bar',
  };

  const result = createStorybookOverride({
    ...storybookDefaultProject,
    settings,
  });

  expect(result?.settings).toStrictEqual(settings);
});

test('allows passing env', () => {
  const env: Parameters<OverrideCreator>[0]['env'] = {
    browser: true,
  };

  const result = createStorybookOverride({
    ...storybookDefaultProject,
    env,
  });

  expect(result?.env).toStrictEqual(env);
});

test('allows passing excludedFiles', () => {
  const excludedFiles = ['src/test.foo'];

  const result = createStorybookOverride({
    ...storybookDefaultProject,
    excludedFiles,
  });

  expect(result?.excludedFiles).toStrictEqual(excludedFiles);
});

test('allows passing extends', () => {
  const customExtends = ['some-config'];

  const result = createStorybookOverride({
    ...storybookDefaultProject,
    extends: customExtends,
  });

  expect(customExtends.every(e => result?.extends?.includes(e))).toBe(true);
});

test('allows passing globals', () => {
  const globals = {
    foo: true,
  };

  const result = createStorybookOverride({
    ...storybookDefaultProject,
    globals,
  });

  expect(result?.globals).toStrictEqual(globals);
});

test('allows passing plugins', () => {
  const plugins = ['a', 'b', 'c'];

  const result = createStorybookOverride({
    ...storybookDefaultProject,
    plugins,
  });

  expect(plugins.every(plugin => result?.plugins?.includes(plugin))).toBe(true);
});

test('allows passing overrides', () => {
  const overrides: Parameters<OverrideCreator>[0]['overrides'] = [
    {
      files: ['src/test.foo'],
      rules: {
        foo: 'warn',
      },
    },
  ];

  const result = createStorybookOverride({
    ...storybookDefaultProject,
    overrides,
  });

  expect(result?.overrides).toStrictEqual(overrides);
});

describe('override snapshots', () => {
  test('without storybook', () => {
    expect(createStorybookOverride(defaultProject)).toMatchSnapshot();
  });

  test('with react', () => {
    expect(
      createStorybookOverride({
        ...storybookDefaultProject,
        react: {
          ...storybookDefaultProject.react,
          hasReact: true,
        },
      })
    ).toMatchSnapshot();
  });

  test('with typescript', () => {
    expect(
      createStorybookOverride({
        ...storybookDefaultProject,
        typescript: {
          ...storybookDefaultProject.typescript,
          hasTypeScript: true,
        },
      })
    ).toMatchSnapshot();
  });

  test('with storybook-testing-lib', () => {
    expect(
      createStorybookOverride({
        ...storybookDefaultProject,
        storybook: {
          ...storybookDefaultProject.storybook,
          hasStorybookTestingLibrary: true,
        },
      })
    ).toMatchSnapshot();
  });
});
