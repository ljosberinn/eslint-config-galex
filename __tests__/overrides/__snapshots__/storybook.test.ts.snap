// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`override snapshots with react 1`] = `
{
  "env": undefined,
  "excludedFiles": undefined,
  "extends": undefined,
  "files": [
    "*.stories.*",
    "*.story.*",
  ],
  "globals": undefined,
  "overrideType": "eslint-config-galex/storybook",
  "overrides": undefined,
  "parserOptions": undefined,
  "plugins": [
    "storybook",
  ],
  "rules": {
    "import/no-anonymous-default-export": "off",
    "import/no-default-export": "off",
    "react/function-component-definition": "off",
    "storybook/csf-component": "error",
    "storybook/default-exports": "off",
    "storybook/hierarchy-separator": "warn",
    "storybook/meta-inline-properties": "warn",
    "storybook/no-redundant-story-name": "warn",
    "storybook/no-stories-of": "error",
    "storybook/no-title-property-in-meta": "warn",
    "storybook/prefer-pascal-case": "warn",
    "storybook/story-exports": "warn",
    "storybook/use-storybook-expect": "error",
    "storybook/use-storybook-testing-library": "error",
  },
  "settings": undefined,
}
`;

exports[`override snapshots with storybook-testing-lib 1`] = `
{
  "env": undefined,
  "excludedFiles": undefined,
  "extends": undefined,
  "files": [
    "*.stories.*",
    "*.story.*",
  ],
  "globals": undefined,
  "overrideType": "eslint-config-galex/storybook",
  "overrides": undefined,
  "parserOptions": undefined,
  "plugins": [
    "storybook",
  ],
  "rules": {
    "import/no-anonymous-default-export": "off",
    "import/no-default-export": "off",
    "storybook/await-interactions": "error",
    "storybook/context-in-play-function": "error",
    "storybook/csf-component": "error",
    "storybook/default-exports": "off",
    "storybook/hierarchy-separator": "warn",
    "storybook/meta-inline-properties": "warn",
    "storybook/no-redundant-story-name": "warn",
    "storybook/no-stories-of": "error",
    "storybook/no-title-property-in-meta": "warn",
    "storybook/prefer-pascal-case": "warn",
    "storybook/story-exports": "warn",
    "storybook/use-storybook-expect": "error",
    "storybook/use-storybook-testing-library": "error",
  },
  "settings": undefined,
}
`;

exports[`override snapshots with typescript 1`] = `
{
  "env": undefined,
  "excludedFiles": undefined,
  "extends": undefined,
  "files": [
    "*.stories.*",
    "*.story.*",
  ],
  "globals": undefined,
  "overrideType": "eslint-config-galex/storybook",
  "overrides": undefined,
  "parserOptions": undefined,
  "plugins": [
    "storybook",
  ],
  "rules": {
    "@typescript-eslint/explicit-module-boundary-types": "off",
    "@typescript-eslint/no-empty-function": "off",
    "import/no-anonymous-default-export": "off",
    "import/no-default-export": "off",
    "storybook/csf-component": "error",
    "storybook/default-exports": "off",
    "storybook/hierarchy-separator": "warn",
    "storybook/meta-inline-properties": "warn",
    "storybook/no-redundant-story-name": "warn",
    "storybook/no-stories-of": "error",
    "storybook/no-title-property-in-meta": "warn",
    "storybook/prefer-pascal-case": "warn",
    "storybook/story-exports": "warn",
    "storybook/use-storybook-expect": "error",
    "storybook/use-storybook-testing-library": "error",
  },
  "settings": undefined,
}
`;

exports[`override snapshots without storybook 1`] = `null`;
